socorro/external/postgresql/signature_first_date.py

```diff
@@ -35,3 +37,51 @@ def get(self, **kwargs):
             'hits': signatures,
             'total': len(signatures)
         }
+
+    def post(self, **kwargs):
+        filters = [
+            ('signature', None, str),
+            ('first_report', None, datetime.datetime),
+            ('first_build', None, str),
+        ]
+        params = external_common.parse_arguments(filters, kwargs)
+        sql_upsert = """
+        WITH
+        update_signature AS (
+            UPDATE signatures
+            SET
+                first_report = LEAST(
+                    %(first_report)s,
+                    (SELECT first_report FROM signatures WHERE signature = %(signature)s)
+                ),
+                first_build = LEAST(
+                    %(first_build)s,
+                    (SELECT first_build FROM signatures WHERE signature = %(signature)s)
+                )
+            WHERE
+                signature = %(signature)s
+            RETURNING 1
+        ),
+        insert_signature AS (
+            INSERT INTO
+                signatures
+                (signature, first_report, first_build)
+            SELECT
+                %(signature)s AS signature,
+                %(first_report)s AS first_report,
+                %(first_build)s AS first_build
+            WHERE NOT EXISTS (
+                SELECT *
+                FROM signatures
+                WHERE
+                    signature = %(signature)s
+                LIMIT 1
+            )
+            RETURNING 2
+        )
+        SELECT * FROM update_signature
+        UNION ALL
+        SELECT * FROM insert_signature
+        """
```

Comment:
I've never actually seen the SQL for a manual upsert before, this is trippy.

Reply:
This is similar to other upsert code in Socorro. I don't understand exactly why the `UNION ALL` thing works the way it does where it tries update and if that doesn't do anything, then does the insert. The postgres docs didn't enlighten me at all on that account.

Reply:
I think the `UNION ALL` does nothing except execute both select statements, which evaluates the queries that populate those two tables. It's the `WHERE NOT EXISTS` bit that only runs the insert for signatures that don't exist yet, and the `WHERE` in the `UPDATE` bit that avoids updating signatures that don't exist. The magic of the `UNION ALL` is that it allows both statements to execute in one query, making it a real upsert instead of faking it with two separate queries.

Reply:
Ahh... ok. I forgot about the `WHERE NOT EXISTS` part. I couldn't figure out why it seemed like only one executed based on the final return codes. That makes soooo much sense now.